src/hb/shape_wasm.rs

```diff
@@ -0,0 +1,728 @@
+
+    caller.data_mut().buffer = ret;
+
+    1
+}
+
+// ===========
+// structs used into WASM
+// ===========
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+enum PointType {
+    MoveTo,
+    LineTo,
+    QuadraticTo,
+    CubicTo,
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+struct CGlyphOutlinePoint {
+    x: f32,
+    y: f32,
+    pointtype: PointType,
+}
+
+unsafe impl bytemuck::Zeroable for CGlyphOutlinePoint {}
+unsafe impl bytemuck::Pod for CGlyphOutlinePoint {}
+
+#[derive(Default)]
+struct GlyphOutline {
+    points: alloc::vec::Vec<CGlyphOutlinePoint>,
+    contours: alloc::vec::Vec<u32>,
+}
+
+impl ttf_parser::OutlineBuilder for GlyphOutline {
+    fn move_to(&mut self, x: f32, y: f32) {
+        self.points.push(CGlyphOutlinePoint {
+            x,
+            y,
+            pointtype: PointType::MoveTo,
+        })
+    }
+
+    fn line_to(&mut self, x: f32, y: f32) {
+        self.points.push(CGlyphOutlinePoint {
+            x,
+            y,
+            pointtype: PointType::LineTo,
+        })
+    }
+
+    fn quad_to(&mut self, x1: f32, y1: f32, x: f32, y: f32) {
+        self.points.push(CGlyphOutlinePoint {
+            x: x1,
+            y: y1,
+            pointtype: PointType::QuadraticTo,
+        });
+        self.points.push(CGlyphOutlinePoint {
+            x,
+            y,
+            pointtype: PointType::QuadraticTo,
+        });
+    }
+
+    fn curve_to(&mut self, x1: f32, y1: f32, x2: f32, y2: f32, x: f32, y: f32) {
+        self.points.push(CGlyphOutlinePoint {
+            x: x1,
+            y: y1,
+            pointtype: PointType::CubicTo,
+        });
+        self.points.push(CGlyphOutlinePoint {
+            x: x2,
+            y: y2,
+            pointtype: PointType::CubicTo,
+        });
+        self.points.push(CGlyphOutlinePoint {
+            x,
+            y,
+            pointtype: PointType::CubicTo,
+        });
+    }
+
+    fn close(&mut self) {
+        // def len not len -1
+        self.contours.push(self.points.len() as u32)
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+#[repr(C)]
+struct CGlyphOutline {
+    n_points: u32,
+    points: u32, // pointer
+    n_contours: u32,
+    contours: u32, // pointer
+}
+
+unsafe impl bytemuck::Zeroable for CGlyphOutline {}
+unsafe impl bytemuck::Pod for CGlyphOutline {}
+
+/// Some data provided by ~~Harfbuzz~~. rustybuzz
+#[derive(Debug, Clone, Copy)]
+#[repr(C)]
+pub struct Blob {
+    /// Length of the blob in bytes
+    pub length: u32,
+    /// A raw pointer to the contents
+    pub data: u32, // *mut u8
+}
+
+unsafe impl bytemuck::Zeroable for Blob {}
+unsafe impl bytemuck::Pod for Blob {}
+
+// using rustybuzz types instead of custom types
+#[derive(Debug, Clone, Copy)]
+#[repr(C)]
+struct CBufferContents {
+    length: u32,
+    info: u32,
+    position: u32,
+}
+
+unsafe impl bytemuck::Zeroable for CBufferContents {}
+unsafe impl bytemuck::Pod for CBufferContents {}
+
+#[cfg(test)]
+mod wasm_tests {
```

Comment:
Can you move tests to the tests directory?

Reply:
Done. But I am not sure if that's better for the Ruqaa test, as it now doesn't distinguish between the wasm code failing and faulty output.

Reply:
How so? Can you elaborate?

Reply:
It is not a very well formed thought, but the wasm shaper can fail in two ways (categories of ways, rather). Either `shape_with_wasm` returns with bad output, or it returns `None`, because of some wasm failure.  Since the tests in the tests folder only test the public API instead of private functions, they *always* return, as they fall back to the regular shaper. So one might not be able to tell if the wasm module crashed or not.